Clerk Elements (Beta) - Initial docs (#890)
clerk/clerk-docs: docs/elements/reference/sign-in.mdx

---
description: Reference documentation for Clerk Elements sign-in components.
---

# Sign-in components

> The following components are used when creating sign-in flows. They are imported from `@clerk/elements/sign-in`. It's recommended to import them all under the `SignIn.` namespace to make discovery easier and reduce naming conflicts with other components throughout your application. The code snippets on this page assume you have imported the components this way.

Reviewer: ```suggestion
The following components are used when creating sign-in flows. They are imported from `@clerk/elements/sign-in`. We recommend importing them all under the `SignIn.` namespace to make discovery easier and reduce naming conflicts with other components throughout your application. The code snippets on this page assume you have imported the components this way.
```

> ```tsx filename="Anatomy"

Author: Yeah, so this is intended to showcase all of the possible steps in the sign in flow, like Radix outlines: https://www.radix-ui.com/primitives/docs/components/accordion#anatomy

import * as SignIn from '@clerk/elements/sign-in'

export default function SignInPage() {
  return (
    <SignIn.Root>
      <SignIn.Step name="start" />
      <SignIn.Step name="verifications" />
      <SignIn.Step name="choose-strategy" />
      <SignIn.Step name="forgot-password" />
      <SignIn.Step name="reset-password" />
    </SignIn.Root>
  )
}
```

## `<Root>`

> The root sign-in component. Sets up providers and state management for the sign-in flow. Must wrap all other sign-in components. `<Root>` will validate your built sign-in flow to ensure the implementation is correct based on instance settings and best practices.

Author: Yes, it will error in development.

  </SignIn.SupportedStrategy>
  <p>or</p>
  <SignIn.SupportedStrategy name="google">
    Sign in with Google
  </SignIn.SupportedStrategy>
</SignIn.Step>
```

### `<Step name="reset-password">`

If a user has requested a password reset and verified their identity, they will be navigated to `reset-password`. A password field should be rendered in this step to allow the user to set a new password.

#### Example

```tsx filename="page.tsx"
<SignIn.Step name="reset-password">
  <Clerk.Field name="password">
    <Clerk.Label>New password</Clerk.Label>
    <Clerk.Input />
    <Clerk.FieldError />
  </Clerk.Field>
  <Clerk.Field name="confirmPassword">
    <Clerk.Label>Confirm password</Clerk.Label>
    <Clerk.Input />
    <Clerk.FieldError />
  </Clerk.Field>
  <SignIn.Action submit>Update password</SignIn.Action>
</SignIn.Step>
```

## `<Strategy>`

Conditionally render its children depending on the authentication strategy that needs to be verified. Does not render any markup on its own.

### Properties

| Name | Type     | Description                                                       |
| ---- | -------- | ----------------------------------------------------------------- |

> | `name` | `'saml' \| 'ticket' \| 'password' \| 'passkey' \| 'phone_code' \| 'email_code' \| 'web3_metamask_signature' \| 'reset_password_email_code' \| 'reset_password_phone_code' \| 'email_link' \| 'totp' \| 'backup_code' \| 'oauth' \| 'web3'` \| [OAuthStrategy](https://github.com/clerk/javascript/blob/956d8792fefe9d6a89022f1e938149b25503ec7f/packages/types/src/strategies.ts#L15) | The name of the strategy for which its children will be rendered. |

Reviewer 2: we should wrap OAuthStrategy in a code snippet as well, we can wrap the entire line like this:

```suggestion
| `name` | <code>'saml' \| 'ticket' \| 'password' \| 'passkey' \| 'phone_code' \| 'email_code' \| 'web3_metamask_signature' \| 'reset_password_email_code' \| 'reset_password_phone_code' \| 'email_link' \| 'totp' \| 'backup_code' \| 'oauth' \| 'web3' \| [OAuthStrategy](https://github.com/clerk/javascript/blob/956d8792fefe9d6a89022f1e938149b25503ec7f/packages/types/src/strategies.ts#L15)</code> | The name of the strategy for which its children will be rendered. |
```

https://github.com/clerk/clerk-docs/assets/98043211/c425c15b-1450-48e6-a746-4f4312674899

  <SignIn.Strategy name="password">
    <Clerk.Field name="password">
      <Clerk.Label>Password</Clerk.Label>
      <Clerk.Input />
      <Clerk.FieldError />
    </Clerk.Field>
    <SignIn.Action submit>Sign in</SignIn.Action>
    <SignIn.Action navigate="forgot-password">Forgot password?</SignIn.Action>
  </SignIn.Strategy>
</SignIn.Step>
```

#### Resend

```tsx filename="page.tsx" {9-16}
<SignIn.Step name="verifications">
  <SignIn.Strategy name="email_code">
    <Clerk.Field name="code">
      <Clerk.Label>Code</Clerk.Label>
      <Clerk.Input />
      <Clerk.FieldError />
    </Clerk.Field>
    <SignIn.Action submit>Verify</SignIn.Action>
    <SignIn.Action
      resend
      fallback={({ resendableAfter }) => (
        <p>Resend code in {resendableAfter} second(s)</p>
      )}
    >
      Resend code
    </SignIn.Action>
  </SignIn.Strategy>
</SignIn.Step>
```

## `<SafeIdentifier />`

Render an identifier that has been provided by the user during a sign-in attempt. Renders a `string` (or empty string if it can't find an identifier).

### Examples

#### Basic

```tsx filename="page.tsx"
<SignIn.Strategy name="email_code">
  <h1>Check your email</h1>

>   <p>We've sent a code to <SignIn.SafeIdentifier />.</p>

Reviewer: ```suggestion
  <p>We sent a code to <SignIn.SafeIdentifier />.</p>
```
